### ISSUE TITLE
[OOM] Huge memory usage in MoE

### ISSUE THREAD

**Issue by hilbertmeng** [AI-Hypercomputer/maxtext#1828]
Hi,  I encountered Out of Memory Error when I train a 19B Llama MoE model (~190B total parameters and 19B activated parameters) on v5p-256, which uses 400G memory, significantly exceeding 95G memory limit. 

Here is the key model config modification on the [base.yaml ](https://github.com/hilbertmeng/maxtext/blob/main/MaxText/configs/base.yml).
```
# batch_size and sequence length
max_target_length = 4096
per_device_batch_size = 8.0

# Dense config
base_emb_dim = 5120
base_num_query_heads = 40
base_num_kv_heads = 40
base_mlp_dim = 12288
base_num_decoder_layers = 60
head_dim = 128
vocab_size = 151936

# MoE config
base_mlp_dim = 6144 # base_mlp_dim refers to hidden_dim of each expert 
num_experts_per_tok = 2
num_experts = 32
megablox = True
sparse_matmul = True
shared_experts = 0
```
To analyze the memory usage, I only reduced `base_num_decoder_layers` config from 60 to 6, so that it can run without OOM and I profiled the memory usage below. 
<img width="1410" alt="Image" src="https://github.com/user-attachments/assets/b2fbef04-0cde-4e25-8856-c112cd6d28ac" />

I found the memory usage constantly increases during backpropagation (Program Order 1500-4000).  A further analysis reveals that [tgmm](https://github.com/hilbertmeng/maxtext/blob/main/MaxText/kernels/megablox/gmm.py#L619) kernel saves each layer's expert weight of shape (32, 6144, 5120), consuming 1.9G memory, so three `linear` (w0, w1 and wo) count ~6G. Thus for a 60-layer model, `tgmm` consumes 60x6=~360G, close to the observed 400G memory usage.  Here I provide a typical piece of expert memory usage in `tgmm`.  

```
Size:
1920.00 MiB
Unpadded Size:
1920.00 MiB
Shape (and minor-to-major order):
bf16[32,6144,5120]{2,1,0:T(8,128)(2,1)}
Tf Op Name::
jit(train_step)/jit(main)/transpose(jvp(Transformer))/decoder/checkpoint/rematted_computation/layers_5/sub_0/moe/moe.sparse_matmul/jit(shmap_body)/jit(tgmm)/pallas_call
Allocation Type:
Temporary 
```
Is this a memory leakage of `tgmm` or is there something wrong in my training config? How can I fix it?

**Comment by Skylion007** [comment]
@hilbertmeng I had a similar issue when just trying to load DeepSeekV3 config on GPUs (after fixing a dtype error in the mask for cudnn_attention for MLA).  The fact that it also tries to allocate a similar amount of memory on GPUs (causing an immedaite OOM), makes me think there is an XLA flag missing we need to use it or something.

```bash
SLICES="${SLURM_NTASKS}"          # total processes
ICI_TP=1
ICI_FSDP=8                        # each process sees exactly 1 local device
ICI_DP=1
ICI_EP=8
DCN_EP=16
DCN_TP=1
DCN_FSDP="${SLICES}"              # shard model across all processes
DCN_DP=1
NUM_SLICES="${SLICES}"

# ---- Launch inside the container; ensure flax is installed for THIS python ----
srun --export=ALL \
     --container-image="${NVME_IMAGE}" \
     --container-mounts="/mnt:/mnt" \
     bash -lc '
       set -euo pipefail

       cd /mnt/sharefs/users/runner/git/maxtext
       # Make sure pip belongs to this python
       python3 -m pip --version >/dev/null 2>&1 || true

       pip install --no-cache -U flax

       #export NVTE_FUSED_ATTN=0
       # Optional: print interpreter path to confirm we are inside the container
       echo "Python: $(command -v python3)"
       python3 -c "import sys; print(sys.version); import flax, jax; print('flax', flax.__version__, 'jax', jax.__version__)"

       # Run MaxText (gpu_multiprocess)
       python3 -m MaxText.train MaxText/configs/base.yml \
         hardware=gpu_multiprocess \
         base_output_directory="'$OUT'" \
         run_name=matmul_pre_training \
         dataset_type=synthetic \
         model_name=deepseek3-671b \
         steps=5 \
         per_device_batch_size=1 \
         max_target_length=1024 \
         enable_checkpointing=false \
         async_checkpointing=false \
         tokenizer_type=huggingface \
         tokenizer_path=deepseek-ai/DeepSeek-V3 \
         attention=cudnn_flash_te \
	 dtype=bfloat16 \
         weight_dtype=bfloat16 \
         megablox=False \
         sparse_matmul=False \
         # ---- ICI (must multiply to 1 for gpu_multiprocess) ----
         ici_expert_parallelism='$ICI_EP' \
         ici_tensor_parallelism='"$ICI_TP"' \
         ici_fsdp_parallelism='"$ICI_FSDP"' \
         ici_data_parallelism='"$ICI_DP"' \
         # ---- DCN (must multiply to num_slices) ----
         dcn_expert_parallelism='"$DCN_EP"' \
         dcn_tensor_parallelism='"$DCN_TP"' \
         dcn_fsdp_parallelism='"$DCN_FSDP"' \
         dcn_data_parallelism='"$DCN_DP"' \
	 remat_policy=full \
	 expert_shard_attention_option=EP_AS_CONTEXT
     '

echo "=== Done. Logs and metrics in: $OUT ==="
```

If there is some issue with the config let me know. I have H200s so this level of FSDP should work on 128 or mode nodes, but sadly it OOMs.
